File: app/src/main/java/com/sortedqueue/programmercreek/fragments/AlgorithmIndexFragment.java
package com.sortedqueue.programmercreek.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.firebase.database.DatabaseError;
import com.sortedqueue.programmercreek.R;
import com.sortedqueue.programmercreek.activity.AlgorithmListActivity;
import com.sortedqueue.programmercreek.adapter.AlgorithmsRecyclerAdapter;
import com.sortedqueue.programmercreek.adapter.CustomProgramRecyclerViewAdapter;
import com.sortedqueue.programmercreek.database.AlgorithmsIndex;
import com.sortedqueue.programmercreek.database.firebase.FirebaseDatabaseHandler;
import com.sortedqueue.programmercreek.interfaces.AlgorithmNavigationListener;

import java.util.ArrayList;

import butterknife.Bind;
import butterknife.ButterKnife;

/**
 * Created by Alok Omkar on 2017-03-17.
 */

public class AlgorithmIndexFragment extends Fragment implements FirebaseDatabaseHandler.GetAllAlgorithmsListener, CustomProgramRecyclerViewAdapter.AdapterClickListner {

    @Bind(R.id.modulesRecyclerView)
    RecyclerView programListRecyclerView;
    private AlgorithmsRecyclerAdapter algorithmsRecyclerAdapter;

    private static AlgorithmIndexFragment indexFragment;
    public static AlgorithmIndexFragment getInstance() {
        if( indexFragment == null ) {
            indexFragment = new AlgorithmIndexFragment();
        }
        return indexFragment;
    }

    private AlgorithmNavigationListener algorithmNavigationListener;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if( context instanceof AlgorithmNavigationListener ) {
            algorithmNavigationListener = (AlgorithmNavigationListener) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        algorithmNavigationListener = null;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_module, container, false);
        ButterKnife.bind(this, view);
        fetchAlgorithmsList();
        return view;
    }

    private void fetchAlgorithmsList() {
        new FirebaseDatabaseHandler(getContext()).getAllAlgorithmIndex( this );
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ButterKnife.unbind(this);
    }

    @Override
    public void onSuccess(ArrayList<AlgorithmsIndex> algorithmsIndexArrayList) {
        programListRecyclerView.setLayoutManager( new LinearLayoutManager(getContext()) );
        algorithmsRecyclerAdapter = new AlgorithmsRecyclerAdapter( getContext(), this, algorithmsIndexArrayList );
        programListRecyclerView.setAdapter( algorithmsRecyclerAdapter );
    }

    @Override
    public void onError(DatabaseError databaseError) {

    }

    @Override
    public void onItemClick(int position) {
        algorithmNavigationListener.loadAlgorithmFragment(algorithmsRecyclerAdapter.getItemAtPosition( position ));
    }


}

File: app/src/main/java/com/sortedqueue/programmercreek/fragments/AlgorithmIndexFragment.kt
package com.sortedqueue.programmercreek.fragments

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.firebase.database.DatabaseError
import com.sortedqueue.programmercreek.R
import com.sortedqueue.programmercreek.adapter.AlgorithmsRecyclerAdapter
import com.sortedqueue.programmercreek.adapter.CustomProgramRecyclerViewAdapter
import com.sortedqueue.programmercreek.database.AlgorithmsIndex
import com.sortedqueue.programmercreek.database.firebase.FirebaseDatabaseHandler
import com.sortedqueue.programmercreek.interfaces.AlgorithmNavigationListener
import kotlinx.android.synthetic.main.fragment_algorithm_index.*

import java.util.ArrayList




/**
 * Created by Alok Omkar on 2017-03-17.
 */

class AlgorithmIndexFragment : Fragment(), FirebaseDatabaseHandler.GetAllAlgorithmsListener, CustomProgramRecyclerViewAdapter.AdapterClickListner {

    private var algorithmsRecyclerAdapter: AlgorithmsRecyclerAdapter? = null

    private var algorithmNavigationListener: AlgorithmNavigationListener? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is AlgorithmNavigationListener) {
            algorithmNavigationListener = context
        }
    }

    override fun onDetach() {
        super.onDetach()
        algorithmNavigationListener = null
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater!!.inflate(R.layout.fragment_algorithm_index, container, false)


        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        fetchAlgorithmsList()
    }

    private fun fetchAlgorithmsList() {
        FirebaseDatabaseHandler(context).getAllAlgorithmIndex(this)
    }

    override fun onDestroyView() {
        super.onDestroyView()

    }

    override fun onSuccess(algorithmsIndexArrayList: ArrayList<AlgorithmsIndex>) {
        programListRecyclerView!!.layoutManager = LinearLayoutManager(context)
        algorithmsRecyclerAdapter = AlgorithmsRecyclerAdapter(context, this, algorithmsIndexArrayList)
        programListRecyclerView!!.adapter = algorithmsRecyclerAdapter
    }

    override fun onError(databaseError: DatabaseError?) {

    }

    override fun onItemClick(position: Int) {
        algorithmNavigationListener!!.loadAlgorithmFragment(algorithmsRecyclerAdapter!!.getItemAtPosition(position))
    }

    companion object {

        @SuppressLint("StaticFieldLeak")
        private var indexFragment: AlgorithmIndexFragment? = null
        val instance: AlgorithmIndexFragment
            get() {
                if (indexFragment == null) {
                    indexFragment = AlgorithmIndexFragment()
                }
                return indexFragment!!
            }
    }


}
